### PR TITLE
Update gravity, create messages in the web interface when necessary (running with cron), see AdminLTE PR 2565

### PR DESCRIPTION
**Pull request by jpgpi250**
Signed-off-by: jpgpi250 <jpgpi250@gmail.com>

This PR will generate messages in the web interface (tools / diagnosis) when gravity is updated non-interactively (cron). If the message already exists, only the timestamp of the message will be updated, this to prevent duplicates. Messages are added when gravity.sh exits (exit 1), only if this can happen non-interactively. If the exit can only happen when running the pihole command with options (from terminal) no messages will be added.

**an additional PR for the web interface is required, will update this PR, as soon I have created it.**

## Thank you for your contribution to the Pi-hole Community!

Please read the comments below to help us consider your Pull Request.

We are all volunteers and completing the process outlined will help us review your commits quicker.

**Please make sure you**

 1. Base your code and PRs against the repositories developmental branch.
 2. [Sign Off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits as we enforce the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
 3. [Sign](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all your commits as they must have verified signatures
 4. File a pull request for any change that requires changes to [our documentation](https://docs.pi-hole.net/) at our [documentation repo](https://github.com/pi-hole/docs) 

---

**What does this PR aim to accomplish?:**

This PR will generate messages in the web interface (tools / diagnosis) when gravity is updated non-interactively (cron). If the message already exists, only the timestamp of the message will be updated, this to prevent duplicates. Messages are added when gravity.sh exits (exit 1), only if this can happen non-interactively. If the exit can only happen when running the pihole command with options (from terminal) no messages will be added.

**How does this PR accomplish the above?:**

added  create_diagnosis_message()
and lines to excecute the code when required (non-interactive exit 1)

**an additional PR for the web interface is required, will update this PR, as soon I have created it.**

Tested with test script and gravity.sh, running interactively and with cron.
The easiest way to test this is to add two lines to the end of gravity.sh:
```create_diagnosis_message "from $0."
echo -e "  ${TICK} from $0."
```

you could also use a test script:
```
#!/usr/bin/env bash

FTLDBfile="/etc/pihole/pihole-FTL.db"

# Create pihole diagnosis message
create_diagnosis_message() {
  declare result=""
  # Only create the message if running non-interactive (cron)
  if pstree -apl "$(pgrep -n cron)" | grep -q "$(basename "$0")"; then
    # prevent duplicates, change the timestamp.
    result=$(pihole-FTL sqlite3  "${FTLDBfile}" "SELECT * FROM message WHERE message LIKE '%$1%';")
    if [ -z "${result}" ]; then
      sudo pihole-FTL sqlite3 "${FTLDBfile}" "INSERT OR IGNORE into message ( timestamp, type, message, blob1 ) values ('$(date +%s)', 'SCRIPT', '$1', '$(base64 <<< "$(basename "$0")")');"
    else 
      sudo pihole-FTL sqlite3 "${FTLDBfile}" "UPDATE message SET timestamp = '$(date +%s)';"
    fi
  fi
}

create_diagnosis_message "from $0."
sleep 5
exit 0
```

you than need to create a cron job to run test.sh or gravity.sh (copy the line from /etc/cron.d/pihole and change time and day)

**Link documentation PRs if any are needed to support this PR:**

To ensure nothing goes wrong in the future, the SCRIPT type must be reserved in https://raw.githubusercontent.com/pi-hole/FTL/master/src/database/message-table.c
---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

### CODE REVIEW

**Comment by jpgpi250** [comment]
- fixed the smoke-test (trailing spaces)
- this was fist mentioned here by DanSchaper, [here](https://discourse.pi-hole.net/t/list-handling-after-recent-pi-hole-update-is-exhausting-nearly-all-resources/62006/91).
-  DL6ER commented, [here](https://discourse.pi-hole.net/t/list-handling-after-recent-pi-hole-update-is-exhausting-nearly-all-resources/62006/107)., whit the remark "probably we have to do this"
- an additional PR is needed, will report the PR# as soon as I have created it.
- I have discovered an additional problem with gravity, error, but gravity continues, will attempt to fix this in an other PR (one thing at the time)...
-

File: gravity.sh [hunk 1]
```diff
@@ -82,6 +87,21 @@ if [[ -r "${piholeDir}/pihole.conf" ]]; then
...
+create_diagnosis_message() {
+  declare result=""
+  # Only create the message if running non-interactive (cron)
+  if pstree -apl "$(pgrep -n cron)" | grep -q "$(basename "$0")"; then
+    # prevent duplicates, change the timestamp.
+    result=$(pihole-FTL sqlite3  "${FTLDBfile}" "SELECT * FROM message WHERE message LIKE '%$1%';")
+    if [ -z "${result}" ]; then
+      sudo pihole-FTL sqlite3 "${FTLDBfile}" "INSERT OR IGNORE into message ( timestamp, type, message, blob1 ) values ('$(date +%s)', 'SCRIPT', '$1', '$(base64 <<< "$(basename "$0")")');"
+    else
+      sudo pihole-FTL sqlite3 "${FTLDBfile}" "UPDATE message SET timestamp = '$(date +%s)';"
```
Comment (DL6ER) [inline gravity.sh:100]: The command as is updates the timestamp of *all* messages in the table (there are no `WHERE ...`) limitations. You can also use `sqlite3` internal functions to set the timestamp, avoiding the call to `date`:

```suggestion
      sudo pihole-FTL sqlite3 "${FTLDBfile}" "UPDATE message SET timestamp = (cast(strftime('%s', 'now') as int)) WHERE message = '$1' AND type = 'GRAVITY';"
```

**Review by DL6ER — CHANGES_REQUESTED** [review]
I have a few suggestions for improvement. Please test them carefully. I'm not able to test this myself right now so consider them as suggestions without guarantee that they work 100% as they are written here.

File: gravity.sh [hunk 2]
```diff
@@ -82,6 +87,21 @@ if [[ -r "${piholeDir}/pihole.conf" ]]; then
   echo -e "  ${COL_LIGHT_RED}Ignoring overrides specified within pihole.conf! ${COL_NC}"
 fi
 
+# Create pihole diagnosis message
+create_diagnosis_message() {
+  declare result=""
+  # Only create the message if running non-interactive (cron)
+  if pstree -apl "$(pgrep -n cron)" | grep -q "$(basename "$0")"; then
+    # prevent duplicates, change the timestamp.
+    result=$(pihole-FTL sqlite3  "${FTLDBfile}" "SELECT * FROM message WHERE message LIKE '%$1%';")
```
[the remark this answers is not in the document]
Comment (jpgpi250) [reply gravity.sh:96]: 'type = 'SCRIPT, see AdminLTE PR 2564. The goal is to make this possible from any script, including personal scripts. That is why the scriptname is included in the final message ($0). Can agree to change LIKE into IS, whil change that, but can't signoff due to the type mismatch.

File: gravity.sh (hunk 1 above)
Comment (jpgpi250) [reply gravity.sh:100]: 'type = 'SCRIPT, see AdminLTE PR 2564. The goal is to make this possible from any script, including personal scripts. That is why the scriptname is included in the final message ($0). Can agree to change $(date +%s) into the suggestion, will change that, but can't signoff due to the type mismatch.

File: gravity.sh [hunk 3]
```diff
@@ -82,6 +87,21 @@ if [[ -r "${piholeDir}/pihole.conf" ]]; then
...
+    if [ -z "${result}" ]; then
+      sudo pihole-FTL sqlite3 "${FTLDBfile}" "INSERT OR IGNORE into message ( timestamp, type, message, blob1 ) values ('$(date +%s)', 'SCRIPT', '$1', '$(base64 <<< "$(basename "$0")")');"
```
[the remark this answers is not in the document]
Comment (jpgpi250) [reply gravity.sh:98]: 'type = 'SCRIPT, see AdminLTE PR 2564. The goal is to make this possible from any script, including personal scripts. That is why the scriptname is included in the final message ($0). In order to use type 'SCRIPT' **AND** show the script name in the message, I entered the message in blob1. The message is the the same, possibly shorter as the 'echo -e' message, displayed when running pihole -g interactively. I tried entering the message into blob1, using alternate methods, but failed, using BASE64 is the only option that works (at least that I know of).

**Comment by jpgpi250** [comment]
I accidently created this PR with Master as target, doesn't look like I can change this. Is it possible to change this?

**Comment by rdwebdesign** [comment]
Yes.

Click on `Edit` button (the same one used to edit the PR title).
It will allow you to change the target branch, right bellow the title.

**Comment by jpgpi250** [comment]
I followed the steps from https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits to squash the commits
Than force pushed the gravity diagnosis branch. 
Is this correct? 
If NOT:
- Do I need to start over again? I will if I have to, after all conversations have been resolved.
- Could somebody please tell me how to squash, step by step.
Running on windows, Using github desktop to clone, create new branch, edit (with notepad++, create PR
have installed git for windows (latest 64 bit
If I right click on the repository I can open a command prompt and issue git commands in a 'cmd'
image

apologies for my ignorance.

**Comment by jpgpi250** [comment]
@DL6ER 
unless somebody else is silently opposed to this going forward, there are two reasons I can identify for this to be in a deadlock. If there is somebody strongly opposed to this PR, just say so, I'll close them both.
1. 'SCRIPT' type:
in order to keep this code usable from multiple scripts, yet make it clear it is a different category, the following suggestions:
- BASH
- CORE (from pi-hole core - GitHub)
2. base64 encoding:
In order to keep the structure of the message, line 1 identifies the script, line 2 contains the actual message, see above screenshot, I needed to put the the script name ($0) in a blob (which annoys you, I accept the argument - unreadable without extra tools). This could solved by a database change, simply add an additional TEXT field. This would have nearly no impact on the database size (very few messages) and would not require any changes in the admin LTE code for the already existing types, messages.js only refers to what it uses (change required for the additional code - this PR).

If we could agree upon a middel way here (pick a type and add a TEXT field to the message table, I would redo the PR, this to avoid the squash problems I've been having with GitHub desktop.

**Comment by dschaper** [comment]
I haven't really looked at this but I don't see where the discussion of why this is needed has happened. 

The PR message has a really nice "What" description but I don't see "Why" it's needed, can you expand on that?

You'll also need to rebase your changes on the latest `development` branch code, the release and hotfix commits should not be a part of this PR.

**Comment by jfb-pihole** [comment]
"weekly automatic updates..."

To be clear, you are referring only to cron driven gravity rebuilds and not user-scripted unattended auto updates of the Pi-hole software?

**Comment by jpgpi250** [comment]
cron driven gravity, but as explained in my replies to DL6ER, this could be used by any script to report a problem (pihole -d for example), hence the scrip name ($0) in the message.

**Comment by dschaper** [comment]
> When users are running pi-hole on auto pilot, weekly automatic updates, and never look at the logs (if they even know the logs exists), this PR will add warning messages to the web interface, informing them something has gone wrong, gravity hasn't been updated.

Okay, that sounds logical and something to consider.



> but [this](https://github.com/pi-hole/pi-hole/issues/5244) issue shows that in some cases, the web interface makes it look like gravity has run to completion successfully , even if it hasn't.

That is a bug that needs to be fixed. The web interface should accurately reflect the current state.



> but as explained in my replies to DL6ER, this could be used by any script to report a problem (pihole -d for example), hence the scrip name ($0) in the message.

That part I'm going to veto. I do not want unchecked user scripts running under the user and privileges of `pihole`. That's a massive security consideration that I don't want to go near until some time after version 6 at the earliest.

> rebase: having some troubles with [GitHub desktop squash](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits), when making change upon change. I will close this PR and create a new one, if it is decided to go forward with this change, with the changes agreed upon.

Rebasing and squashing are two separate concepts in `git`. Rebase takes the existing commits and applies them to a different HEAD (branch). That's what needs to happen here, take the existing commits and base them on the `development` branch of the `pihole` repo. No need to squash anything until you have successfully rebased the existing commit history. After that you can look at squashing your PR's commits (which is essentially rebasing with a commit in the current history. I've always found an interactive rebase to be the best way to squash.)

**Pull request closed without merging**